reflred/backgroundfield.py

```diff
@@ -0,0 +1,256 @@
+r"""
+Background field subtraction
+"""
+from __future__ import division
+
+import numpy as np
+import scipy.optimize as so
+import matplotlib.pyplot as plt
+from copy import copy
+
+#from dataflow.lib.uncertainty import Uncertainty as U, interp
+#from .refldata import ReflData
+
+# BackgroundFieldData class returns results of background field fitting
+
+class BackgroundFieldData(object):
+    def __init__(self, p, dp2, s, ds2, dsp, chisq):
+        self.p = p
+        self.dp2 = dp2
+        self.s = s
+        self.ds2 = ds2
+        self.dsp = dsp
+        self.chisq = chisq
+    def get_metadata(self):
+        return {
+            "epsD": self.p,
+            "epsD_var": self.dp2,
+            "scale": self.s,
+            "scale_var": self.ds2,
+            "scale_epsD_covar": self.dsp,
+            "fit_chi-squared": self.chisq
+        }
+    def get_plottable(self):
+        return self.get_metadata()
+
+
+def background_reservoir(ai, epsD, af, epssi, Fd, scale):
+    """
+    Calculates background from a reservoir surrounded by Si
+    """
+    A = np.exp(-epsD * (1. / np.sin(ai) + 1. / np.sin(af)))
```

Comment:
This will cause problems if ai or af are zero.  

A simple solution is to place the whole function in `errstate` so that divide by zero produces inf or nan as appropriate:
```
with np.errstate(divide='ignore'):
    ...
```

In this case, the result would be e^-inf, which evaluates to A=0.

Similarly for line 43, which will evaluate to 1 if ai=0, or 0 if af=0 or NaN if both are zero, and line 45, which also produces 0/0 if both ai and af are zero.

Not sure what happens to values that are near but not equal to zero.  For example, with line 43 has numerator `(1-A) = (1-exp(...)) = -(exp(...) - 1` is better expressed using the `expm1` function for maximum accuracy. Perhaps not needed in this case, but in cases with inaccurate numerator and denominator, then you may end up with effectively random values.

This is all moot if ai and af are never zero.  If we exclude the "specular" condition ai=af=0, then perhaps the with-statement will be enough.

Reply:
This calculation is not intended to be used at very small angles (there's always beam bleed-in). Points around zero should be excluded from the calculation. I have implemented the with statement, however.